### ISSUE TITLE
build-all.sh generated larger executable and it cost much longer to convert to OBJ

### ISSUE THREAD

**Issue by huantingchen** [IfcOpenShell/IfcOpenShell#137]
Hi, just grabbed the latest commit in Master to build a static binary using \nix\build-all.sh. The build was successful but I got a larger executable (IfcConvert was 51.8MB) compared to the 0.5.0-preview1 one (downloaded from official website) which was only 23.8 MB. Besdies, testing w/ a sample model, my build took over 6 mins to convert to OBJ while 0.5.0-preview1 only took 75 secs. I didn't modify the source code. What did I do wrong? Or is there any other way to build a standalone executable of IfcConvert? Please advise. Thanks!

**Comment by aothms** [comment]
> but I got a larger executable

This is to be expected. The build script produces binaries that have full unicode support with ICU compiled in, which is (yet another) big dependency.

> took over 6 mins to convert 

This is not to be expected. Did you test some more models? Did you experience similar performance issues on other (sufficiently different) files? The code base has changed significantly, also dependencies have been updated. Time increase could also come from the fact that by now more elements are supported.

Actually there have been various improvements wrt performance (e.g #47 #82). Would be good to have an understanding of what is going on. Hope you can share some more details.

**Comment by huantingchen** [comment]
I am managing to find more IFC models to give it a go. Can share some benchmarks later. However, are you able to tag 0.5.0 version so that I can also build it on my machine to compare the performance? Thanks.

**Comment by huantingchen** [comment]
Some benchmarks here. Looks like both file size and conversion time increase drastically. I don't feel comfortable with it since my application would do the conversion on server side. Any thoughts?

==version 0.5.0-preview-1==
| Model                            | IFC size      | OBJ size      | conversion time |
| -------------------------------- | ------------- | ------------- | --------------- |
| A: Revit structure model         | 11.8 MB       | 19.4 MB       | 39 secs         |
| B: Revit architectural model (1) | 20.8 MB       | 65.6 MB       | 89 secs         |
| C: Revit architectural model (2) | 28.3 MB       | 66.0 MB       | 74 secs         |
| D: Revit MEP model               | 184.4 MB      | 526.1 MB      | 368 secs        |
| E: Tekla model w/ rebar          | 12.1 MB       | 1.8 GB        | 170 secs        |
| F: Tekla model w/o rebar         | 12.1 MB       | 16.3 MB       | 36 secs         |

==latest commit build==
| Model                            | IFC size      | OBJ size      | conversion time |
| -------------------------------- | ------------- | ------------- | --------------- |
| A: Revit structure model         | 11.8 MB       | 29.1 MB       | 230 secs        |
| B: Revit architectural model (1) | 20.8 MB       | 102.6 MB      | 588 secs        |
| C: Revit architectural model (2) | 28.3 MB       | 110 MB        | 110 secs        |
| F: Tekla model w/o rebar         | 12.1 MB       | 25.4 MB       | 154 secs        |

**Comment by huantingchen** [comment]
Some benchmarks here. Looks like both file size and conversion time increase drastically. I don't feel comfortable with it since my application would do the conversion on server side. Any thoughts?

version 0.5.0-preview-1

| Model                            | IFC size      | OBJ size      | conversion time |
| -------------------------------- | ------------- | ------------- | --------------- |
| A: Revit structure model         | 11.8 MB       | 19.4 MB       | 39 secs         |
| B: Revit architectural model (1) | 20.8 MB       | 65.6 MB       | 89 secs         |
| C: Revit architectural model (2) | 28.3 MB       | 66.0 MB       | 74 secs         |
| D: Revit MEP model               | 184.4 MB      | 526.1 MB      | 368 secs        |
| E: Tekla model w/ rebar          | 12.1 MB       | 1.8 GB        | 170 secs        |
| F: Tekla model w/o rebar         | 12.1 MB       | 16.3 MB       | 36 secs         |

latest commit build

| Model                            | IFC size      | OBJ size      | conversion time |
| -------------------------------- | ------------- | ------------- | --------------- |
| A: Revit structure model         | 11.8 MB       | 29.1 MB       | 230 secs        |
| B: Revit architectural model (1) | 20.8 MB       | 102.6 MB      | 588 secs        |
| C: Revit architectural model (2) | 28.3 MB       | 110 MB        | 110 secs        |
| F: Tekla model w/o rebar         | 12.1 MB       | 25.4 MB       | 154 secs        |

**Comment by Stinkfist0** [comment]
In OBJ (and DAE) floats/doubles are now written more precision which probably explains the change in output size. I have plans to add command-line switch for specifying the output precision in the near future.

**Comment by huantingchen** [comment]
Correct. The precision now is double precision which is 15 decimal places. The old version only output 4 decimal places. However, the conversion time still takes very long as shown above.

I built on Ubuntu using build-all.sh in \nix. Isn't this script used to build the binary that I downloaded from the website?

**Comment by aothms** [comment]
> I built on Ubuntu using build-all.sh in \nix. Isn't this script used to build the binary that I downloaded from the website?

It was, but like the source code, the build script continuously evolves as well.

> Also, on MSVC I can remember noticing that enabling more aggressive inlining (/Ob2) had a tremendous impact on IfcConvert's speed to process files. 

It does, there are tons of indirections and function calls in the OCCT codebase, if they are not inlined, higher level optimizations cannot do much good.

I will dive into this tomorrow, one thing to try would be whether `BUILD_CFG=Release ./build-all.sh` makes any difference. After the binaries have been built on the website, the nix build script has been modified to match the windows build script. There the default is `RelWithDebInfo`, which is a meaningful default for a development environment. Off the top of my head I am not sure whether that boils down to `-O1`, `-O2` or `-O3` on Ubuntu (@stinkfist, wouldn't be so sure its `-O3`, on Windows I must admit the performance difference is minimal) and whether `-g` comes with a noticeable performance penalty.

So tl;dr you can try with `BUILD_CFG=Release`, but I will have a look myself tomorrow, so feel free to wait for that.

**Comment by mcukierman** [comment]
Maybe it's also worth to have in mind the other performance issue:
https://sourceforge.net/p/ifcopenshell/bugs/32/

**Comment by huantingchen** [comment]
I've tried `BUILD_CFG=Release ./build-all.sh`. But it took much much longer. For example, my model C took 6 mins......

**Comment by aothms** [comment]
Yeah `BUILD_CFG=Release` doesn't cut it either (or is even worse...). I tested on the Schependomlaan model [1]. `Release` also doesn't imply `-O3` which apparently is really necessary to get performant IfcOpenShell builds from gcc.

cmd | processing time
--- | --- 
`./build-all.sh` | 4 minutes 58 seconds
`BUILD_CFG=Release ./build-all.sh` | 5 minutes 52 seconds
`CFLAGS=-O3 CXXFLAGS=-O3 BUILD_CFG=Release ./build-all.sh` | 1 minute 28 seconds

We probably need to incorporate this into the build script itself. Thanks for bringing this to our attention, I wasn't aware the differences were that astronomical.

[1] https://github.com/openBIMstandards/DataSetSchependomlaan/blob/master/Design%20model%20IFC/IFC%20Schependomlaan.ifc

**Comment by aothms** [comment]
> Maybe it's also worth to have in mind the other performance issue:
> https://sourceforge.net/p/ifcopenshell/bugs/32/

@mcukierman Many boolean op issues have been addressed in OCCT 6.9. I will have another look at the Digital 210 king model and report back as it is a prime example of the bad performance of the old binary boolean op module.

**Comment by Stinkfist0** [comment]
The ENABLE_BUILD_OPTIMIZATIONS can be used to easily enable this.

**Comment by huantingchen** [comment]
Thanks @aothms . I used the following command to build and it looked better. CFLAGS=-O3 CXXFLAGS=-O3 BUILD_CFG=Release ./build-all.sh

Wondering how many IFC entities have been added since 0.5.0 for now. I changed to previous precision so the produced files seemed to be okay. However, the conversion time was still a bit longer compared to 0.5.0.
Some benchmarks below using the same set of test models:


| Model                            | IFC size      | OBJ size      | conversion time |
| -------------------------------- | ------------- | ------------- | --------------- |
| A: Revit structure model         | 11.8 MB       | 19.5 MB       | 51 secs         |
| B: Revit architectural model (1) | 20.8 MB       | 65.7 MB       | 156 secs        |
| C: Revit architectural model (2) | 28.3 MB       | 67.0 MB       | 115 secs        |
| F: Tekla model w/o rebar         | 12.1 MB       | 16.4 MB       | 39 secs         |

**Comment by aothms** [comment]
Can you post a link to one of the models? I would like to have a look how this compares to the other platforms and os'es.

**Comment by huantingchen** [comment]
My OS is ubuntu-14.04.4-desktop-amd64
Please find my test models from here.
Model A: https://drive.google.com/file/d/0B2k_rLbfTrj2VDNkQ2VpX09wSlU/view?usp=sharing
Model B: https://drive.google.com/file/d/0B2k_rLbfTrj2MEd2bGROY1JxMWs/view?usp=sharing
Model C: https://drive.google.com/file/d/0B2k_rLbfTrj2NGYyNFdud3RxMUk/view?usp=sharing
Model F: https://drive.google.com/file/d/0B2k_rLbfTrj2c3NJYW92LXItdk0/view?usp=sharing

**Comment by aothms** [comment]
Thanks. For Model F, I get a 10 times longer processing time, but otherwise the trend is consistent on Windows. Still not completely ruled out that the difference is in the compiler optimization though.

model | ubuntu-14.04 master | win preview | win  master
--- | --- | --- | ---
test.ifc | 51 | 58.6524999142 | 69.9735000134
model_20140104.ifc | 156 | 121.726500034 | 183.065000057
test20160925.ifc | 115 | 84.1694998741 | 140.163499951
tekla.ifc | 39 | 388.234500051 | 618.130500078

I compared the contents in terms of IfcRepresentationItems for the smallest time increase and the largest.

test.ifc || tekla.ifc ||
--- | --- | --- | --- 
IfcCartesianPoint | 31403 |                IfcMappedItem | 34447
IfcFace | 24716 |                          IfcCartesianPoint | 11015
IfcFaceOuterBound | 24716 |                IfcCartesianTransformationOperator3D | 5998
IfcPolyLoop | 24716 |                      IfcTrimmedCurve | 4148
IfcAxis2Placement3D | 7941 |               IfcStyledItem | 3342
IfcPolyline | 2862 |                       IfcAxis2Placement3D | 3272
IfcExtrudedAreaSolid | 2634 |              IfcCompositeCurveSegment | 2980
IfcAxis2Placement2D | 2622 |               IfcSweptDiskSolid | 2739
IfcCompositeCurveSegment | 918 |           IfcFace | 2496
IfcMappedItem | 787 |                      IfcPolyLoop | 1276
IfcStyledItem | 516 |                      IfcFaceOuterBound | 1264
IfcClosedShell | 355 |                     IfcCompositeCurve | 1214
IfcFacetedBrep | 355 |                     IfcClosedShell | 515
IfcPlane | 345 |                           IfcFacetedBrep | 515
IfcBooleanClippingResult | 345 |           IfcLine | 461
IfcPolygonalBoundedHalfSpace | 334 |       IfcDirection | 315
IfcCircle | 217 |                          IfcCircle | 254
IfcTrimmedCurve | 217 |                    IfcVector | 188
IfcCompositeCurve | 149 |                  IfcBooleanResult | 178
IfcDirection | 54 |                        IfcExtrudedAreaSolid | 165
IfcGeometricSet | 35 |                     IfcPolyline | 67
IfcGeometricCurveSet | 12 |                IfcAxis2Placement2D | 22
IfcHalfSpaceSolid | 11 |                   IfcHalfSpaceSolid | 19
IfcConnectedFaceSet | 4 |                  IfcPlane | 19
IfcFaceBasedSurfaceModel | 4 |             IfcBooleanClippingResult | 16
IfcCartesianTransformationOperator3D | 1 | IfcFaceBound | 12 
                                     | |   IfcGeometricCurveSet | 7

Seems that tekla.ifc features a lot of `IfcSweptDiskSolid`s. Nothing substantial changed there, so it might be caused by some differences in the dependencies. I will check what the results look like when I set the dependencies to the time of the previous release. OCE-0.14 and no ICU if not mistaken.

**Comment by huantingchen** [comment]
@aothms Thanks for looking into this. For Model F, I excluded IfcReinforcingBar. That was why I got much faster.

**Comment by aothms** [comment]
> I excluded IfcReinforcingBar

Ok thanks for clarifying.

So, in order to get a better understanding of what happened I built IfcOpenShell using the current build script for some 150 revisions in the past and let IfcConvert churn on the "Revit structure model". If you zoom in on the image you can see the individual commit ids.

![performance](https://cloud.githubusercontent.com/assets/1096535/19265377/3739dcf4-8fa5-11e6-82e8-3091e803d197.png)

There is a large spike when testing caching behaviour: https://github.com/IfcOpenShell/IfcOpenShell/commit/8ecc126. Other than that, the results are clearly bimodal. Unfortunately there is a large void in which either IfcOpenShell didn't compile on Ubuntu, or IfcConvert failed in processing the file. Somewhere in this void the performance got worse. Between https://github.com/IfcOpenShell/IfcOpenShell/commit/191c029 and https://github.com/IfcOpenShell/IfcOpenShell/commit/01269e5. I am going to have to repeat this on Windows as (and switch to topological ordering for the commits) to hopefully pinpoint the offending commits.

**Comment by huantingchen** [comment]
> Between 191c029 and 01269e5
So something went wrong here. I would also like to give older commits (before 191c029) a go.

**Comment by aothms** [comment]
Update: I did look into this on Windows, which didn't deliver additional insights.

My next attempt was to process all elements individually and see how performance has changed on the individual items. No significant change there. Perhaps something pertaining to the caching then. 

![splitted_performance](https://cloud.githubusercontent.com/assets/1096535/19639631/8448bc2a-99d9-11e6-925d-639c5fcd33cd.png)

I have to let this rest for a while due to other priorities, but I will leave the issue open for further investigation.

**Comment by huantingchen** [comment]
@aothms Since we've worked w/ Ubuntu for a long time, probably just build it from older commits for now. Thanks again for looking into this.

**Comment by Moult** [comment]
I tried to download the files to get a little status update 6 years later, but didn't have access to the files (I've requested access). @aothms do you know any other insights that might impact this bug in the timeframe?

**Comment by aothms** [comment]
I don't think that what we're seeing here is really relevant still for the current code base.

One tip for future readers. With `IfcConvert -vvvv --log-format=json ...` you'll get detailed performance reporting on the individual processing of elements in case somebody wonders about performance anomalies in the future.
